File: CSES-sheet/Graph/Flight-Discount-task-1195/code.cpp
#include <bits/stdc++.h>
using namespace std;

using ll = long long;

const ll INF = 1e18;

struct Edge {
    int u, v, wt;
};

vector<ll> dijkstra(int start, int n, vector<vector<pair<int, int>>> &adj) {
    vector<ll> dist(n + 1, INF);

    priority_queue<pair<ll, int>, vector<pair<ll, int>>, greater<pair<ll, int>>> pq;

    dist[start] = 0;
    pq.push({0, start});

    while (!pq.empty()) {
        auto [currwt, u] = pq.top();
        pq.pop();

        if (currwt != dist[u])
            continue;

        for (auto &[v, wt] : adj[u]) {
            if (currwt + wt < dist[v]) {
                dist[v] = currwt + wt;
                pq.push({dist[v], v});
            }
        }
    }

    return dist;
}

void solve() {
    int n, m;
    cin >> n >> m;

    vector<vector<pair<int, int>>> adj(n + 1), revadj(n + 1);
    vector<Edge> edges;

    for (int i = 0; i < m; i++) {
        int u, v, wt;
        cin >> u >> v >> wt;

        adj[u].push_back({v, wt});
        revadj[v].push_back({u, wt});

        edges.push_back({u, v, wt});
    }

    vector<ll> dist1 = dijkstra(1, n, adj);
    vector<ll> distN = dijkstra(n, n, revadj);

    ll cheapest = INF;

    for (auto &e : edges) {
        if (dist1[e.u] != INF && distN[e.v] != INF) {
            ll cost = dist1[e.u] + e.wt / 2 + distN[e.v];
            cheapest = min(cheapest, cost);
        }
    }

    cout << cheapest << '\n';
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    solve();

    return 0;
}

File: CSES-sheet/Graph/Flight-Discount-task-1195/code.java
import java.io.*;
import java.util.*;

public class Demo {

   // static char[][] adj;
  ///  static List<Integer> cycle;

    static char[] dir = { 'U', 'R', 'D', 'L' };
    static int[] drow = { -1, 0, 1, 0 };
    static int[] dcol = { 0, 1, 0, -1 };

    static List<int[]> edges;
    static long INF=(long)1e18;

    public static long[] dijkstras(int start,int n,int m,List<List<int[]>> adj){
        long[] distance=new long[n+1];

        Arrays.fill(distance, INF);

        distance[start]=0;

        PriorityQueue<long[]> pq =new PriorityQueue<>((a,b)->Long.compare(a[1], b[1]));

        pq.add(new long[]{start,0,-1});

        while(!pq.isEmpty()){
            long[] curr=pq.poll();

            int u=(int)curr[0];
            long currwt=curr[1];
            int par =(int)curr[1];

            List<int[]> edges=adj.get(u);
            for(int[] e:edges){
             
                int v=e[0];

                if(v==par)continue;
                int wt=e[1];

                if(distance[u]<INF && currwt+wt <  distance[v]){
                    distance[v]=currwt+wt;
                    pq.add(new long[]{v,distance[v],u});
                }
            }
        }

        return distance;
    }

   

    public static void solve(Scanner sc) {

        int n = sc.nextInt();
        int m = sc.nextInt();
       
    
         List<List<int[]>> adj=new ArrayList<>();
         List<List<int[]>> revadj=new ArrayList<>();
        edges=new ArrayList<>();
        
        for(int i=0;i<=n;i++){
            adj.add(new ArrayList<>());
            revadj.add(new ArrayList<>());
        }



        for(int i=0;i<m;i++){
            int u=sc.nextInt();
            int v=sc.nextInt();
            int wt=sc.nextInt();

         
           adj.get(u).add(new int[]{v,wt});
           revadj.get(v).add(new int[]{u,wt});

           edges.add(new int[]{u,v,wt});

        }
        
        long[] distance_from_1=dijkstras(1, n, m,adj);
        long[] distance_from_n=dijkstras(n, n, m,revadj);
        
        long cheapest=INF;

        for(int[]e:edges){

             int u=e[0];
             int v=e[1];
             int wt=e[2];
             
             if(distance_from_1[u]<INF && distance_from_n[v]<INF){

                long res=(long)Math.floor((double)wt/2.0);
                cheapest=Math.min(cheapest,distance_from_1[u]+res+distance_from_n[v]);
             }

        }

        System.out.println(cheapest);


    


    }

    public static void main(String[] args) throws Exception {
        Scanner sc = new Scanner(System.in);
        solve(sc);
        sc.close();
    }
}
